### PR TITLE
Manage return focus to activator for Secondary actions

### PR DESCRIPTION
**Pull request by oksanashopify**
### WHY are these changes introduced?

**_Resolves:_** https://github.com/Shopify/polaris/issues/8105

_**TL;DR:**_ There is a focus management issue prevalent across the Admin. Many of the Page secondary actions in the Admin - e.g. Import, Export, Duplicate, etc. - open a modal, and currently there is no easy way to [return focus to the modal activator](https://www.w3.org/WAI/ARIA/apg/patterns/dialogmodal/#keyboard-interaction-7) when the activator is the secondary action of the Page component that doesn't support it.

### WHAT is this pull request doing?

- [commit 5151eaa0](https://github.com/Shopify/polaris/pull/8722/commits/5151eaa074d4718d58b7d05be3338bd6d3dc2b95): added activatorRef to SecondaryAction to manage return focus to activator.
Video how it works in Admin:

https://user-images.githubusercontent.com/104942025/226687638-970f1ec5-7609-4ed9-b16d-04fd397cff3e.mov

- [commit 8bfa834](https://github.com/Shopify/polaris/pull/8722/commits/8bfa834c5d593165091c22ee7d95240e22965328): added activatorRef to Grouped action to manage return focus to activator
Video how it works in Admin:

https://user-images.githubusercontent.com/104942025/226687935-e22c7cad-5446-43f4-a3e2-0983a6ad6786.mov

### How to 🎩
Run Polaris and[ corresponding web](https://github.com/Shopify/web/pull/87000) branches locally, please follow these instructions: [ Working with Polaris](https://web.docs.shopify.io/docs/guides/working-with-polaris).
[Spin URL](https://shop1.shopify.oa-add-fer-to-secondary-actions-new.oksana-azarova.us.spin.dev/admin/staff) for testing implementation for Product Detail "Duplicate" and "More Actions" -> "Fake App action" buttons

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

Note: There are only Page `secondaryActions `and `actionGroups` provided in a playground, as using these actions and Modal requires special web utilities. So it is better to use a [real web branch](https://github.com/Shopify/web/pull/87000) to test implementation.
<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  const duplicateActivatorRef = useRef<HTMLElement | null>(null);
  const moreActionsActivatorRef = useRef<HTMLElement | null>(null);
  return (
    <>
      <Page
        breadcrumbs={[{content: 'Products', url: '#'}]}
        title="3/4 inch Leather pet collar"
        titleMetadata={<Badge status="success">Paid</Badge>}
        subtitle="Perfect for any pet"
        compactTitle
        primaryAction={{content: 'Save', disabled: true}}
        secondaryActions={[
          {
            content: 'Duplicate',
            accessibilityLabel: 'Secondary action label',
            onAction: () => alert('Duplicate action'),
            activatorRef: duplicateActivatorRef,
          },
        ]}
        actionGroups={[
          {
            title: 'More actions',
            actions: [
              {
                content: 'Share on Facebook',
                accessibilityLabel: 'Individual action label',
                onAction: () => alert('Share on Facebook action'),
              },
            ],
            groupActivatorRef: moreActionsActivatorRef,
          },
        ]}
      />
    </>
  );
}

```

</details>

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/components/ActionMenu/components/SecondaryAction/SecondaryAction.tsx [hunk 1]
```diff
@@ -9,6 +9,7 @@ import styles from './SecondaryAction.scss';
 
 interface SecondaryAction extends ButtonProps {
   helpText?: React.ReactNode;
+  activatorRef?: LegacyRef<HTMLSpanElement> | undefined;
```
Comment (oksanashopify) [inline polaris-react/src/components/ActionMenu/components/SecondaryAction/SecondaryAction.tsx:12]: What would be the best place on https://polaris.shopify.com/ to reflect in documentation changes represented in this PR?
I was looking for [Page section](https://polaris.shopify.com/components/layout-and-structure/page) with `With custom secondary action`, but seems `secondaryActions` has type `any`

File: polaris-react/src/components/ActionMenu/components/MenuGroup/MenuGroup.tsx [hunk 2]
```diff
@@ -22,6 +22,8 @@ export interface MenuGroupProps extends MenuGroupDescriptor {
   getOffsetWidth?(width: number): void;
   /** Collection of sectioned action items */
   sections?: readonly ActionListSection[];
+  /** The element or the RefObject that activates the Modal */
+  groupActivatorRef?: LegacyRef<HTMLSpanElement> | undefined;
```
Comment (oksanashopify) [inline polaris-react/src/components/ActionMenu/components/MenuGroup/MenuGroup.tsx:26]: What would be the best place on https://polaris.shopify.com/ to reflect in documentation changes represented in this PR?
I was looking for [Page section](https://polaris.shopify.com/components/layout-and-structure/page) with With custom secondary action, but seems `actionGroups` has type `MenuGroupDescriptor[]` and my new property was added to interface `ComplexAction`

File: polaris-react/src/types.ts [hunk 3]
```diff
@@ -242,6 +246,8 @@ export interface MenuGroupDescriptor extends BadgeAction {
   onActionAnyItem?: ActionListItemDescriptor['onAction'];
   /** Callback when the menu is clicked */
   onClick?(openActions: () => void): void;
+  /** The element or the RefObject that activates the Modal */
+  groupActivatorRef?: LegacyRef<HTMLSpanElement> | undefined;
```
Comment (tatianau) [inline polaris-react/src/types.ts:250]: This is great and it seems to work for the secondary actions and group actions that are defined at the consumer level _explicitly_ through `secondaryActions`  and `actionGroups`  props. 

It doesn't seem to handle the case with the _implicitly_ collapsed actions, for example, when `Duplicate` action becomes part of `More actions` when there is not enough real estate (smaller screens, higher zoom level, etc.) - in those cases the focus is lost. Is that something we could fix here as well?

https://user-images.githubusercontent.com/15054990/226711770-83804f6c-9dc0-4050-b467-bbebec020ab3.mp4

File: polaris-react/src/components/ActionMenu/components/SecondaryAction/SecondaryAction.tsx [hunk 4]
```diff
@@ -30,9 +32,11 @@ export function SecondaryAction({
   }, [getOffsetWidth]);
 
   const buttonMarkup = (
-    <Button onClick={onAction} {...rest}>
-      {children}
-    </Button>
+    <span ref={activatorRef}>
```
Comment (knames) [inline polaris-react/src/components/ActionMenu/components/SecondaryAction/SecondaryAction.tsx:35]: [this](https://github.com/Shopify/polaris/pull/8701/files#diff-9c8f2ff3b7f4b56e6f5a156e5aa48458d62578bcbd45e8400ba0771760e32ece) resolves having to wrap buttons to include refs

**Comment by knames** [comment]
any interest in slotting this PR behind [mine](https://github.com/Shopify/polaris/pull/8701/files#diff-9c8f2ff3b7f4b56e6f5a156e5aa48458d62578bcbd45e8400ba0771760e32ece)? it does away with wrapping buttons and doesn't require using legacyRef

File: polaris-react/src/components/ActionMenu/components/MenuGroup/MenuGroup.tsx [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import React, {useCallback} from 'react';
+import React, {LegacyRef, useCallback} from 'react';
```
Comment (knames) [inline polaris-react/src/components/ActionMenu/components/MenuGroup/MenuGroup.tsx:1]: could we use forwardRef here in place of LegacyRef?

**Review by chloerice — CHANGES_REQUESTED** [review]
Hey @oksanashopify 👋🏽 As @knames mentioned, there's a [PR](https://github.com/Shopify/polaris/pull/8701) up to add `ref` support across actions holistically. These changes won't be necessary when #8701 ships.

You'll be able to set a `ref` on the action objects passed in to `Page.secondaryActions` and to the objects passed into the `actions` properties of `Page.actionGroups` directly and `Button` will do the rest. SecondaryActionProps extends ButtonProps and spreads `...rest` onto Button, so the `ref` will pass through without needing to add a new property to SecondaryAction itself.

**Comment by oksanashopify** [comment]
Closed in favour of https://github.com/Shopify/polaris/pull/8701
